src/src/PluginStructs/P176_data_struct.cpp

```diff
@@ -0,0 +1,370 @@
+      # ifndef BUILD_NO_DEBUG
+      addLog(LOG_LEVEL_DEBUG, F("Victron: Serial port started"));
+      # endif // ifndef BUILD_NO_DEBUG
+    }
+  }
+
+  return isInitialized();
+}
+
+/*****************************************************
+* plugin_read
+*****************************************************/
+bool P176_data_struct::plugin_read(struct EventStruct *event)           {
+  bool success = false;
+
+  if (isInitialized()) {
+    for (uint8_t i = 0; i < VARS_PER_TASK; ++i) {
+      String key = getTaskValueName(event->TaskIndex, i);
+      key.toLowerCase();
+      String value;
+
+      if (getReceivedValue(key, value)) {
+        int32_t iValue      = 0;
+        const float fFactor = getKeyFactor(key);
+
+        if (validIntFromString(value, iValue)) {
+          UserVar.setFloat(event->TaskIndex, i, iValue * fFactor);
+          success = true;
+        }
+      }
+    }
+  }
+
+  return success;
+}
+
+// In order of most values included
+// mV, mAh
+const char p176_factor_1000[] PROGMEM =
+  "v|v2|v3|vs|vm|vpv|i|i2|i3|il|ce|h1|h2|h3|h6|h7|h8|h15|h16"
+;
+
+// 0.01 kWh, 0.01 V
+const char p176_factor_0_01[] PROGMEM =
+  "h17|h18|h19|h20|h22|ac_out_v|dc_in_v|"
+;
+
+// promille
+const char p176_factor_10[] PROGMEM =
+  "dm|soc|"
+;
+
+// 0.1 A
+const char p176_factor_0_1[] PROGMEM =
+  "ac_out_i|dc_in_i|"
+;
+
+float P176_data_struct::getKeyFactor(const String& key) const {
+  if (key.isEmpty()) { return 1.0f; }
+
+  if (GetCommandCode(key.c_str(), p176_factor_1000) > -1) {
+    return 0.001f;
+  }
+
+  if (GetCommandCode(key.c_str(), p176_factor_0_01) > -1) {
+    return 100.0f;
+  }
+
+  if (GetCommandCode(key.c_str(), p176_factor_10) > -1) {
+    return 0.1f;
+  }
+
+  if (GetCommandCode(key.c_str(), p176_factor_0_1) > -1) {
+    return 10.0f;
+  }
+
+  return 1.0f;
+}
+
+/*****************************************************
+* plugin_ten_per_second
+*****************************************************/
+bool P176_data_struct::plugin_ten_per_second(struct EventStruct *event)           {
+  bool success = false;
+
+  if (isInitialized()) {
+    handleSerial(); // Read and process data
+    success = true;
+  }
+
+  return success;
+}
+
+/*****************************************************
+* plugin_get_config_value
+*****************************************************/
+bool P176_data_struct::plugin_get_config_value(struct EventStruct *event,
+                                               String            & string)           {
+  bool success = false;
+
+  if (isInitialized()) {
+    const String key      = parseString(string, 1);
+    const String decimals = parseString(string, 2); // Optional decimals, default = 2, neg. = trim trailing zeroes
+    String value;
+
+    if (getReceivedValue(key, value)) {
+      int32_t iValue      = 0;
+      const float fFactor = getKeyFactor(key);
+
+      if (!essentiallyEqual(fFactor, 1.0f) && validIntFromString(value, iValue)) {
+        int32_t nrDecimals = 2;
```

Comment:
Shouldn't you also fill in the nr. of decimals in the same function where you get the factor?
Makes the code simpler and you're not loosing decimals for "factor 1000" values.